From b8dc0582d00c838da9834831a18e030c9e254b7b Mon Sep 17 00:00:00 2001
From: Eberhard Beilharz <eb1@sil.org>
Date: Wed, 14 Dec 2011 09:40:23 +0100
Subject: [PATCH 3/3] bug-xamarin-2394: Fix crash when setting Row.Height

When setting both Row.Height and MinimumHeight we need to set the
MinimumHeight first otherwise we will get a crash if the new Height
value is less than the old MinimumHeight value.

Also added a unit test that demonstrates the problem.

Change-Id: I5dabf98c373bafc2f700712e4d1b4c4911ebd911
---
 .../System.Windows.Forms/DataGridView.cs           |   43 +++++++---------
 .../Test/System.Windows.Forms/DataGridViewTest.cs  |   54 ++++++++++++++++++++
 2 files changed, 72 insertions(+), 25 deletions(-)

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
@@ -3123,6 +3123,20 @@
 				InvalidateRow (i);
 		}
 
+		private void UpdateRowHeightInfo (DataGridViewRow row)
+		{
+			DataGridViewRowHeightInfoNeededEventArgs rowInfo =
+				new DataGridViewRowHeightInfoNeededEventArgs (row.Index, row.Height, row.MinimumHeight);
+			OnRowHeightInfoNeeded (rowInfo);
+
+			if (row.Height != rowInfo.Height || row.MinimumHeight != rowInfo.MinimumHeight) {
+				row.MinimumHeight = rowInfo.MinimumHeight;
+				row.Height = rowInfo.Height;
+				OnRowHeightInfoPushed (new DataGridViewRowHeightInfoPushedEventArgs (row.Index, rowInfo.Height,
+													rowInfo.MinimumHeight));
+			}
+		}
+
 		public void UpdateRowHeightInfo (int rowIndex, bool updateToEnd)
 		{
 			if (rowIndex < 0 && updateToEnd)
@@ -3142,33 +3156,12 @@
 
 			if (updateToEnd) {
 				for (int i = rowIndex; i < Rows.Count; i++) {
-					DataGridViewRow row = Rows[i];
-					if (!row.Visible)
-						continue;
-
-					DataGridViewRowHeightInfoNeededEventArgs rowInfo = 
-						new DataGridViewRowHeightInfoNeededEventArgs (row.Index, row.Height, row.MinimumHeight);
-					OnRowHeightInfoNeeded (rowInfo);
-
-					if (row.Height != rowInfo.Height || row.MinimumHeight != rowInfo.MinimumHeight) {
-						row.Height = rowInfo.Height;
-						row.MinimumHeight = rowInfo.MinimumHeight;
-						OnRowHeightInfoPushed (new DataGridViewRowHeightInfoPushedEventArgs (row.Index, rowInfo.Height, 
-														     rowInfo.MinimumHeight));
-					}
+					DataGridViewRow row = Rows [i];
+					if (row.Visible)
+						UpdateRowHeightInfo (row);
 				}
 			} else {
-				DataGridViewRow row = Rows[rowIndex];
-				DataGridViewRowHeightInfoNeededEventArgs rowInfo = 
-					new DataGridViewRowHeightInfoNeededEventArgs (row.Index, row.Height, row.MinimumHeight);
-				OnRowHeightInfoNeeded (rowInfo);
-
-				if (row.Height != rowInfo.Height || row.MinimumHeight != rowInfo.MinimumHeight) {
-					row.Height = rowInfo.Height;
-					row.MinimumHeight = rowInfo.MinimumHeight;
-					OnRowHeightInfoPushed (new DataGridViewRowHeightInfoPushedEventArgs (row.Index, rowInfo.Height, 
-													     rowInfo.MinimumHeight));
-				}
+				UpdateRowHeightInfo (Rows [rowIndex]);
 			}
 		}
 
Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
@@ -1825,6 +1825,60 @@
 			DataGridView gdv = new DataGridView ();
 			Assert.IsNull (gdv.RowTemplate.DataGridView, "#1");
 		}
+
+		[Test]
+		public void RowHeightLessThanOldMinHeightVirtMode ()
+		{
+			using (var dgv = new DataGridView ()) {
+				dgv.VirtualMode = true;
+				dgv.RowCount = 1;
+				dgv.Rows [0].Height = 10;
+				dgv.Rows [0].MinimumHeight = 5;
+				dgv.RowHeightInfoNeeded += (sender, e) => {
+					e.Height = 2;
+					e.MinimumHeight = 2;
+				};
+				dgv.UpdateRowHeightInfo (0, false);
+				Assert.AreEqual (2, dgv.Rows [0].Height);
+				Assert.AreEqual (2, dgv.Rows [0].MinimumHeight);
+			}
+		}
+
+		[Test]
+		[ExpectedException (typeof (ArgumentOutOfRangeException))]
+		public void RowHeightLessThanMinHeightVirtMode ()
+		{
+			using (var dgv = new DataGridView ()) {
+				dgv.VirtualMode = true;
+				dgv.RowCount = 1;
+				dgv.Rows [0].Height = 10;
+				dgv.Rows [0].MinimumHeight = 5;
+				dgv.RowHeightInfoNeeded += (sender, e) => {
+					e.Height = 2;
+					e.MinimumHeight = 5;
+				};
+				// we expect an exception on the next line
+				dgv.UpdateRowHeightInfo (0, false);
+			}
+		}
+
+		[Test]
+		public void MinHeightGreaterThanOldRowHeightVirtMode ()
+		{
+			using (var dgv = new DataGridView ()) {
+				dgv.VirtualMode = true;
+				dgv.RowCount = 1;
+				dgv.Rows [0].Height = 10;
+				dgv.Rows [0].MinimumHeight = 5;
+				dgv.RowHeightInfoNeeded += (sender, e) => {
+					e.MinimumHeight = 30;
+					e.Height = 40;
+				};
+				dgv.UpdateRowHeightInfo (0, false);
+				Assert.AreEqual (40, dgv.Rows [0].Height);
+				Assert.AreEqual (30, dgv.Rows [0].MinimumHeight);
+			}
+		}
 	}
 	
 	[TestFixture]
